### PR TITLE
System metrics allow list

### PR DESCRIPTION
**Pull request by jakub-dzon**
This PR introduces the notion of system metrics allow-list for specifying which system metrics should be exclusively collected from the edge devices.

### CODE REVIEW

File: internal/yggdrasil/yggdrasil.go [hunk 1]
```diff
@@ -233,10 +241,20 @@ func getDeviceMetricsConfiguration(edgeDevice *v1alpha1.EdgeDevice) *models.Metr
 			metricsConfig.System = &models.SystemMetricsConfiguration{
 				Interval: systemMetrics.Interval,
 			}
+
+			allowListSpec := systemMetrics.AllowList
+			if allowListSpec != nil {
+				allowList, err := h.allowLists.GenerateFromConfigMap(ctx, allowListSpec.Name, allowListSpec.Namespace)
```
Comment (pkliczewski) [inline internal/yggdrasil/yggdrasil.go:247]: Should we have some default metrics we would like to collect? We could add a flag to explicitly disable metrics collection.

**Review by pkliczewski — COMMENTED** [review]
/lgtm

File: internal/yggdrasil/yggdrasil.go (hunk 1 above)
Comment (jakub-dzon) [reply internal/yggdrasil/yggdrasil.go:247]: I want to have defaults on the device side.

I thought about allowing users to disable system metrics collection and I will implement it as a separate couple of PRs.

**Review by pkliczewski — APPROVED** [review]

**Review by gciavarrini — APPROVED** [review]
lgtm

**Review by machacekondra — APPROVED** [review]
I would maybe add some documentation, how to create the configmap, so user know what to define there.

**Comment by jakub-dzon** [comment]
@machacekondra - thanks! Done. I have completely forgotten about the docs. 

@eloycoto - could you, please, update docs for workload metrics after this PR is merged?

File: internal/yggdrasil/yggdrasil.go [hunk 2]
```diff
@@ -60,13 +61,15 @@ type Handler struct {
...
+	allowLists             devicemetrics.AllowListGenerator
 }
 
 type keyMapType = map[string]interface{}
 type secretMapType = map[string]keyMapType
 
 func NewYggdrasilHandler(deviceRepository edgedevice.Repository, deploymentRepository edgedeployment.Repository,
-	claimer *storage.Claimer, k8sClient k8sclient.K8sClient, initialNamespace string, recorder record.EventRecorder, registryAuth images.RegistryAuthAPI, metrics metrics.Metrics) *Handler {
+	claimer *storage.Claimer, k8sClient k8sclient.K8sClient, initialNamespace string, recorder record.EventRecorder,
+	registryAuth images.RegistryAuthAPI, metrics metrics.Metrics, allowLists devicemetrics.AllowListGenerator) *Handler {
```
[the remark this answers is not in the document]
Comment (jakub-dzon) [reply internal/yggdrasil/yggdrasil.go:72]: Yeah, maybe a fluent builder would make it a bit clearer...

File: internal/yggdrasil/yggdrasil.go [hunk 3]
```diff
@@ -215,11 +223,11 @@ func (h *Handler) GetDataMessageForDevice(ctx context.Context, params yggdrasil.
...
 }
 
-func getDeviceMetricsConfiguration(edgeDevice *v1alpha1.EdgeDevice) *models.MetricsConfiguration {
+func (h *Handler) getDeviceMetricsConfiguration(ctx context.Context, edgeDevice *v1alpha1.EdgeDevice) (*models.MetricsConfiguration, error) {
 	metricsConfigSpec := edgeDevice.Spec.Metrics
-	if metricsConfigSpec != nil {
-		var metricsConfig models.MetricsConfiguration
+	var metricsConfig models.MetricsConfiguration
 
+	if metricsConfigSpec != nil {
```
[the remark this answers is not in the document]
Comment (jakub-dzon) [reply internal/yggdrasil/yggdrasil.go:230]: :+1:

**Review by eloycoto — APPROVED** [review]

**Pull request merged**